src/SolARFBOWHelper.cpp

```diff
@@ -77,6 +77,218 @@ double SolARFBOWHelper::distanceBoW(const datastructure::BoWFeature& bow1, const
+	const datastructure::BoWFeature::const_iterator v1_end = v1.end();
+	const datastructure::BoWFeature::const_iterator v2_end = v2.end();
+
+	v1_it = v1.begin();
+	v2_it = v2.begin();
+
+	double score = 0;
+
+	while (v1_it != v1_end && v2_it != v2_end)
+	{
+		const auto& vi = v1_it->second;
+		const auto& wi = v2_it->second;
+
+		if (v1_it->first == v2_it->first)
+		{
+			score += sqrt(vi * wi);
+
+			// move v1 and v2 forward
+			++v1_it;
+			++v2_it;
+		}
+		else if (v1_it->first < v2_it->first)
+		{
+			// move v1 forward
+			v1_it = v1.lower_bound(v2_it->first);
+			// v1_it = (first element >= v2_it.id)
+		}
+		else
+		{
+			// move v2 forward
+			v2_it = v2.lower_bound(v1_it->first);
+			// v2_it = (first element >= v1_it.id)
+		}
+	}
+	return score; // already scaled
+}
+
+double SolARFBOWHelper::distanceDotProductBoW(const datastructure::BoWFeature& v1, const datastructure::BoWFeature& v2) {
+	datastructure::BoWFeature::const_iterator v1_it, v2_it;
+	const datastructure::BoWFeature::const_iterator v1_end = v1.end();
+	const datastructure::BoWFeature::const_iterator v2_end = v2.end();
+
+	v1_it = v1.begin();
+	v2_it = v2.begin();
+
+	double score = 0;
+
+	while (v1_it != v1_end && v2_it != v2_end)
+	{
+		const auto& vi = v1_it->second;
+		const auto& wi = v2_it->second;
+
+		if (v1_it->first == v2_it->first)
+		{
+			score += vi * wi;
+
+			// move v1 and v2 forward
+			++v1_it;
+			++v2_it;
+		}
+		else if (v1_it->first < v2_it->first)
+		{
+			// move v1 forward
+			v1_it = v1.lower_bound(v2_it->first);
+			// v1_it = (first element >= v2_it.id)
+		}
+		else
+		{
+			// move v2 forward
+			v2_it = v2.lower_bound(v1_it->first);
+			// v2_it = (first element >= v1_it.id)
+		}
+	}
+
+	return score; // cannot scale
+}
+
+double SolARFBOWHelper::distanceKLSBoW(const datastructure::BoWFeature& v1, const datastructure::BoWFeature& v2) {
+	datastructure::BoWFeature::const_iterator v1_it, v2_it;
+	const datastructure::BoWFeature::const_iterator v1_end = v1.end();
+	const datastructure::BoWFeature::const_iterator v2_end = v2.end();
+
+	v1_it = v1.begin();
+	v2_it = v2.begin();
+
+	double score = 0;
+	const double LOG_EPS = log(DBL_EPSILON);
+	// all the items or v are taken into account
+
+	while (v1_it != v1_end && v2_it != v2_end)
+	{
+		const auto& vi = v1_it->second;
+		const auto& wi = v2_it->second;
+
+		if (v1_it->first == v2_it->first)
+		{
+			if (vi != 0 && wi != 0) score += vi * log(vi / wi);
+
+			// move v1 and v2 forward
+			++v1_it;
+			++v2_it;
+		}
+		else if (v1_it->first < v2_it->first)
+		{
+			// move v1 forward
+			score += vi * (log(vi) - LOG_EPS);
+			++v1_it;
+		}
+		else
+		{
+			// move v2_it forward, do not add any score
+			v2_it = v2.lower_bound(v1_it->first);
+			// v2_it = (first element >= v1_it.id)
+		}
+	}
+
+	// sum rest of items of v
+	for (; v1_it != v1_end; ++v1_it)
+		if (v1_it->second != 0)
+			score += v1_it->second * (log(v1_it->second) - LOG_EPS);
+
+	return score; // cannot be scaled
```

Comment:
here, the score could be negative ? it seems that nothing ensures the KLS score is always positive ?

Reply:
could have negative values, and should adjust the threshold according to data / use case